### PR TITLE
Add "beta" value to `enable-api-fields`

### PR DESCRIPTION
**Pull request by lbernick**
# Changes

Add "beta" value to `enable-api-fields`
This commit continues the implementation plan described in
[TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md)
by adding a "beta" option to the "enable-api-fields" feature flag.

Features moving from "alpha" to "beta" stability should be gated behind the "beta" value of
this feature flag in v1 until they are ready to move to a "stable" stability level.
Maintainers can make exceptions to this policy on a case-by-case basis.
While the Tekton community may decide to make changes to the API versioning process,
features should follow this "graduation" process until the new process, if any,
is reflected in a TEP.

No functional changes.
/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- n/a Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- n/a Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: docs/install.md
Comment (lbernick) [inline docs/install.md:459]: One thing that might be confusing here is that these docs don't distinguish between fields that are in beta because they're newly introduced, and fields that are in beta because they were in v1beta1 but we don't want to stabilize them. Open to suggestions on how best to document this.

File: pkg/apis/config/feature_flags.go
Comment (JeromeJu) [inline pkg/apis/config/feature_flags.go:207]: nit: wondering if we need to include this in the test? As noticed the coverage off.

**Comment by JeromeJu** [comment]
LGTM
Thank you Lee.

**Review by vdemeester — COMMENTED** [review]
I am not really sure I understand that change. I think in [TEP-0119](https://github.com/tektoncd/community/pull/786/files), we are proposing to deprecate/remove this field for v1 right ? If that's the case, then it shouldn't be there in the `v1` types, at all.

The more we use the `enable-api-fields` feature flag, the more confusing it is getting for me. Aside from the all-or-nothing behavior slight problem, my feeling is that we are abusing the feature-flag for a lot of things.
Shouldn't we make `v1` with the field we only want to support instead ? There is going to be some time when we will (**have to**) support `v1` and `v1beta1` and thus users will still be able to use `step.resources` from `v1beta1` object. Having this "it's in the API (so documented in openapi "possibly", …) but only when the `beta` feature-flag is enable is very confusing to me. It's even more confusing to me as we *aim* to get read of it, but as an end-user, I am mostly assuming what would be under the `beta` feature-flag is on the way to be "promoted".

**Comment by lbernick** [comment]
> The more we use the `enable-api-fields` feature flag, the more confusing it is getting for me. Aside from the all-or-nothing behavior slight problem, my feeling is that we are abusing the feature-flag for a lot of things. Shouldn't we make `v1` with the field we only want to support instead ? There is going to be some time when we will (**have to**) support `v1` and `v1beta1` and thus users will still be able to use `step.resources` from `v1beta1` object. Having this "it's in the API (so documented in openapi "possibly", …) but only when the `beta` feature-flag is enable is very confusing to me. It's even more confusing to me as we _aim_ to get read of it, but as an end-user, I am mostly assuming what would be under the `beta` feature-flag is on the way to be "promoted".

TBH I agree, I would also assume that anything in beta is on its way to be promoted rather than deprecated. Maybe we should reopen discussion around maintaining multiple api versions rather than the enable-api-fields feature flag? @dibyom 

For this PR, would it be helpful to have it just add the beta flag and not use it for any fields right now? (especially since compute resources are still under discussion) Or does it not make sense to add dead code?

**Comment by vdemeester** [comment]
> TBH I agree, I would also assume that anything in beta is on its way to be promoted rather than deprecated. Maybe we should reopen discussion around maintaining multiple api versions rather than the enable-api-fields feature flag? @dibyom

I am not sure I understand the "reopen discussion around maintaining multiple api versions". We are/will maintain multiple api versions, this is a *given*. For a given period of time (according to API policies, …), we will support and serve `v1beta1` as well as `v1`. Which one is stored, and how the controller does its magic is an implementation detail, but from a user perspective, we will support both version, we **have to**.

> For this PR, would it be helpful to have it just add the beta flag and not use it for any fields right now? (especially since compute resources are still under discussion) Or does it not make sense to add dead code?

It kinds of makes it dead code, but I think it's fine though, as it provides the "framework" for `beta` fields to be added / promoted later one.

**Comment by vdemeester** [comment]
> Sorry, I should have been more clear-- I meant instead of adding features gated behind an alpha feature flag to a beta CRD, creating an alpha CRD and adding features there, and later on adding them to the beta CRD. This discussion is probably out of scope for this PR though.

Right, now I understand it a bit more 🙃. Indeed this is a bit out of the scope of this PR, and it comes with its "set" of challenges.

File: pkg/apis/config/feature_flags.go
Comment (lbernick) [reply pkg/apis/config/feature_flags.go:207]: Done.

**Comment by lbernick** [comment]
@dibyom mentioned wanting to discuss potential updates to TEP 33 (whether there is value in a "beta" feature flag at all as opposed to alpha and stable) before merging this PR, so I've added it to the next api wg agenda and I think we can put it on hold for now. FYI @vdemeester I think he might have chatted with you about this as well.

/hold

**Comment by lbernick** [comment]
/hold cancel

discussed at api wg-- don't want to block moving some existing features to beta on the larger discussion about better api versioning, so I'm reopening this pr

**Review by jerop — APPROVED** [review]
@lbernick please add more context/details to the commit message, based on the discussions in this PR and the API WG today

**Comment by lbernick** [comment]
Thanks for the feedback @jerop, done!

**Pull request merged**
